### PR TITLE
feat(auth): added auth handler, validating with jwt

### PR DESCRIPTION
**Pull request by slot**
Missy should be secure by default, for that reason I'm adding an authorization handler.

We assume that there is a iam service and that the request header carries a valid auth jwt token. The validation works with a public certificate that is being passed through a file in the file system the service is running in. The path can be configured in .missy.yml. 

The change is fully backwards compatible. The old function service.Handle() and service.HandleFunc() are marked as deprecated.

### CODE REVIEW

File: service/service.go [hunk 1]
```diff
@@ -115,43 +137,133 @@ func (s *Service) Start() {
...
+						log.Printf("Error: %v", err)
+						return
+					}
+				}
+			}
+		}
+
+		if deadline, ok := ctx.Deadline(); ok {
+			secs := (time.Until(deadline) + time.Second/2) / time.Second
+			log.Printf("Shutdown finished %ds before deadline\n", secs)
```
Comment (levrik) [inline service/service.go]: Actually no `\n` needed for `log.Printf`. It's only needed for `fmt.Printf`.

File: service/service.go [hunk 2]
```diff
@@ -115,43 +137,133 @@ func (s *Service) Start() {
...
 	s.Router.HandleFunc("/info", s.infoHandler).Methods("GET")
 
-	http.Handle("/", s.Router)
+	s.ServeMux.Handle("/", s.Router)
+
+	initPublicKey()
 }
 
 // HandleFunc excepts a HanderFunc an converts it to a handler, then registers this handler
+// Deprecated: Developers should use SecureHandleFunc() or UnsecureHandleFunc() explicitly
```
Comment (levrik) [inline service/service.go]: Can we print a warning to console when used?

File: data/marshaller.go [hunk 3]
```diff
@@ -41,20 +42,16 @@ func Marshal(w http.ResponseWriter, r *http.Request, subject interface{}) {
...
-	} else {
-		convertTo = "json"
-		resp, err = json.Marshal(subject)
-		w.Header().Set(httpHeaderContentType, contentTypeJSON)
-		w.Write(resp)
-		return
 	}
-
+	convertTo = "json"
+	resp, err = json.Marshal(subject)
```
Comment (levrik) [inline data/marshaller.go]: Please move that back and remove the return in line 43/44. It prevents errors from being handled.

File: service/handlers.go [hunk 4]
```diff
@@ -1,12 +1,34 @@
...
+func initPublicKey() {
+	pubkeyLocation := config.GetInstance().Authorization.PublicKeyFile
+	pubkeyPEM, err := ioutil.ReadFile(pubkeyLocation)
+	if err != nil {
+		log.Panic("Unable to load public key file for token auth: ", err)
+	}
+
+	pkey, err := jwt.ParseRSAPublicKeyFromPEM(pubkeyPEM)
+	if err != nil {
+		log.Panic("Unable to parse public key for token auth: ", err)
```
Comment (levrik) [inline service/handlers.go]: Why no `log.Fatal` ?

File: service/service.go (hunk 2 above)
Comment (slot) [reply service/service.go]: I would not, spams you logs. No need to refactor now.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 96dba96738238dfe3d0aaf981ff4de2997ea9f00
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 96dba96738238dfe3d0aaf981ff4de2997ea9f00
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 96dba96738238dfe3d0aaf981ff4de2997ea9f00
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 96dba96738238dfe3d0aaf981ff4de2997ea9f00
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 96dba96738238dfe3d0aaf981ff4de2997ea9f00
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

File: service/service.go [hunk 5]
```diff
@@ -167,7 +167,7 @@ func shutdown(s Shutdowner) {
 	ctx, cancel := context.WithTimeout(context.Background(), timeout)
 	defer cancel()
 
-	log.Printf("" +
+	log.Printf(""+
 		"Server shutdown with timeout: %s", timeout)
```
Comment (levrik) [inline service/service.go]: ?

File: data/marshaller_test.go [hunk 6]
```diff
@@ -7,73 +7,99 @@ import (
...
+		request := httptest.NewRequest("GET", "http://missy.local/marshal", nil)
+		request.Header.Set("Accept", tc.Accept)
+		response := httptest.NewRecorder()
+		MarshalWithCode(response, request, tc.Data, http.StatusTeapot)
+		if ct := response.Header().Get(httpHeaderContentType); ct != tc.ExpectedContentType {
+			t.Logf("Accept header has the wrong content type. Expected %s, actual %s", tc.ExpectedContentType, ct)
+			t.Fail()
+		}
+		if response.Code != http.StatusTeapot {
+			t.Log("Marshaller returned not StatusCode 418 I am a Teapot")
```
Comment (levrik) [inline data/marshaller_test.go]: Actually it's defined as `I'm a teapot` in spec.

Comment (slot) [reply data/marshaller_test.go]: nit picker alert

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 96dba96738238dfe3d0aaf981ff4de2997ea9f00
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 89dea0ef20a7c7ed724b59af98a92129032a93cf
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 89dea0ef20a7c7ed724b59af98a92129032a93cf
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 89dea0ef20a7c7ed724b59af98a92129032a93cf
  * Your subject line is longer than 72 characters
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/microdevs/missy

---
This message was auto-generated by https://gitcop.com

File: data/marshaller_test.go (hunk 6 above)
Comment (levrik) [reply data/marshaller_test.go]: You need to live with that fact :D

**Pull request merged**